src/spinnaker.cpp

```diff
@@ -0,0 +1,1049 @@
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_execute_trigger(struct Camera* self_)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerCamera*)self_)->execute_trigger();
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_get_frame(struct Camera* self_,
+                       void* im,
+                       size_t* nbytes,
+                       struct ImageInfo* info)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerCamera*)self_)->get_frame(im, nbytes, info);
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+//
+// Camera implementation
+//
+
+SpinnakerCamera::SpinnakerCamera(Spinnaker::CameraPtr camera)
+  : Camera{ .set = ::spinnakercam_set,
+            .get = ::spinnakercam_get,
+            .get_meta = ::spinnakercam_get_meta,
+            .get_shape = ::spinnakercam_get_shape,
+            .start = ::spinnakercam_start,
+            .stop = ::spinnakercam_stop,
+            .execute_trigger = ::spinnakercam_execute_trigger,
+            .get_frame = ::spinnakercam_get_frame,
+  }
+  , camera_(camera)
+  , last_known_settings_{}
+  , frame_id_(0)
+  , started_(false)
+{
+    // Sometimes the camera is still initialized from a previous run.
+    // Ideally, we would error here, but that can make the camera indefinitely
+    // unusable in Acquire, so log and reset instead.
+    if (camera->IsInitialized()) {
+        LOGE("Camera was already initialized. De-initializing it before "
+             "re-initializing.");
+        camera_->DeInit();
+    }
+    camera->Init();
+
+    // Acquire only supports certain values of some node values, so set these
+    // once on initialization before getting or setting any other node values
+    // which may depend on them.
+    set_enum_node(camera_->ExposureAuto, genicam_off);
+    set_enum_node(camera_->ExposureMode, genicam_timed);
+
+    get(&last_known_settings_);
+}
+
+SpinnakerCamera::~SpinnakerCamera()
+{
+    try {
+        stop();
+        const std::scoped_lock lock(lock_);
+        camera_->DeInit();
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+}
+
+void
+SpinnakerCamera::set(struct CameraProperties* properties)
+{
+    const std::scoped_lock lock(lock_);
+    // Set shape before offset because Spinnaker blocks updates
+    // to offset if the shape is too big.
+    maybe_set_shape(properties->shape);
+    maybe_set_offset(properties->offset);
```

Comment:
There's an order of operations thing here that I probably also have wrong in egrabber.

Depending on how the camera works you may need to set offset to (0,0) before changing the shape so you have the full range of the sensor. Then set the offset to the desired one.

Reply:
I ran into a case like that hence the comment above. But I think you're right that I may be not handling all cases. Will ensure that tests cover those.

Do we assume that shape takes precedence over offset? E.g. what if a user has previously set a non-zero offset, then sets the shape to the same size as the sensor? In that case, I think we should refuse to set the shape and at least log an error before continuing with other properties.

Reply:
It's a bit like we're setting both at the same time from the perspective of the Acquire api, but spinnaker only allows changing one at a time.

The problem, I've run into is when they have set a valid roi with a non-zero offset previously and then want to return to full sensor. The offset (to 0) needs to be applied before the shape change.

On the other hand, if I'm going in the other direction (full sensor to an valid offset roi), the new shape has to be applied before changing the offset.

In both cases, the user has valid settings. It's just that we need to watch the order of operations so things don't fail.

Reply:
That's clear.

But what if they don't have valid settings? I assume we should not change either offset or shape if the combination is invalid? I.e. given Acquire treat these two as coupled the change should happen atomically or not at all (rather than with offset changed, but not shape).